Separates activation functions from fully connected layers and Introduces activation layers

beomyeol opened the pull request:
This pull request removes activation functions from fully connected layers and introduces activation layers. The activation layer applies the activation function specified by the configuration file to its input value.

This pull request includes the following changes
* Changes API of `LayerBase`'s '`backPropagate()` method.
* Introduces `DefaultLayerParameterInitializer` and an empty layer parameter `EMPTY` for layers which are not learnable.
* Changes the type of exceptions related to invalid indices from `RuntimeException` to `IllegalArgumentException`.
* Fixes bugs related to indices in `backPropagateFromTo()`.
* Fixes typo `isForward`.
* Adds test for activation layers.

The change of `LayerBase`'s `backPropagate()` method API includes the removal of `nextParameter` and the introduction of `input`. The introduction of activation layers makes `nextParameter` unnecessary in `backPropagate()`. Also, although the input value is not needed now, it will be needed for computing gradients of some functions such as absolute value and power functions.

This closes #144.

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetwork.java:
```diff
@@ -185,9 +185,14 @@ public void train(final INDArray input, final int label) {
                                     final INDArray[] activations,
                                     final INDArray label) {
+    if (end == 0) {
+      throw new IllegalArgumentException("The ending layer cannot be the first layer: " +
```
I'm not sure if this is correct. We don't have a separate 'input layer', do we? For example, the first layer in the sample conf file is a fully connected layer, which can have errors.

beomyeol replied:
The error that the first layer calculates is not used for generating gradients for the first layer. Even though the first layer is a fully connected layer, the errors needed for generating gradients are calculated by the next layer's backpropagation. Thus, the backpropagation for the first layer is unnecessary.
I'll add more comments to this.

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetwork.java:
```diff
@@ -205,29 +210,29 @@ public void train(final INDArray input, final int label) {
     for (int i = begin; i >= end; --i) {
-      error = layers[i].backPropagate(activations[i + 1], layers[i + 1].getLayerParameter(), error);
-      errors[i - begin] = error;
+      error = layers[i].backPropagate(activations[i], activations[i + 1], error);
```
This changes which layer we're calling `getLayerParameter` on. Originally, we were using the *next* layer's parameters to compute the weight, but now we're using *this* layer's parameters. Could you explain in more detail what was changed?

beomyeol replied:
In original implementation, A learnable layer (e.g. fully connected layer) calculates the errors that are used to generate gradients for training itself. For this computation, the next layer's parameter (weight and bias) is needed. 
This pull request changes that the errors used for generating gradients are calculated by the next layer. This makes the next layer's parameter unnecessary and learnable layers use their own parameters in backpropagation.

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/conf/ActivationLayerConfigurationBuilder.java:
```diff
@@ -0,0 +1,75 @@
+        .bindNamedParameter(LayerConfigurationParameters.NumberOfOutput.class, String.valueOf(numOutput))
+        .bindNamedParameter(LayerConfigurationParameters.ActivationFunction.class, String.valueOf(activationFunction))
+        .bindImplementation(LayerBase.class, ActivationLayer.class)
+        .bindImplementation(LayerParameterInitializer.class, DefaultLayerParameterInitializer.class)
```
Can we make `DefaultLayerParameterInitializer` as the default implementation of `LayerParameterInitializer` and remove this line?

jsjason commented on dolphin-dnn/src/test/java/edu/snu/dolphin/dnn/NeuralNetworkTest.java:
```diff
@@ -119,8 +137,8 @@ public void buildNeuralNetwork() throws InjectionException {
     final INDArray[] activations = neuralNetwork.feedForward(input);
-    assertTrue(Nd4jUtils.equals(activations, expectedActivations, tolerance));
     assertTrue(Nd4jUtils.equals(activations[activations.length - 1], expectedOutput, tolerance));
+    assertTrue(Nd4jUtils.equals(activations, expectedActivations, tolerance));
```
What's this change for?

beomyeol replied:
An array of activations includes the output of a neural network. If the array of activations that the neural network generates is equal to `expectedActivations` within `tolerance`, the next line before change, that checks the equality between the last element of `activations` and `expectedOutput` was totally unnecessary.
However, if the order is reversed, we may easily figure out the cause of test failure cases such as one of activation values is not returned despite the correct output value.

jsjason commented:
@beomyeol The tests succeed but the example finished with an error (although, the training did progress according to the logs).
```
2015-11-19 15:36:56,171 INFO edu.snu.dolphin.dnn.NeuralNetworkREEF.main main | REEF job completed: FAILED(java.lang.NullPointerException: Null ids parameter.)
```
Am I the only one that met this error?

beomyeol commented:
@jsjason I addressed your comments and left some comments about your questions.
Also, I fixed a bug that the backpropagation does not work well in a neural network with 2 layers and introduced `ActivationWithLossLayer`.

`ActivationWithLossLayer` is a wrapper of `ActivationLayer`. Its forward pass is same to that of `ActivationLayer`, but a backward pass is calculating a derivative of the configured loss function with respect to the output value and the expected output value of the layer. The cross-entropy loss function is included in this pull request. We can add other loss functions later.

The `ActivationWithLossLayer` is needed for supporting other loss functions, and for the consistency of which layer (itself or **the next layer**) calculates errors to generate gradients for training the layer.

beomyeol commented:
@jsjason I did not experience the error that you said. Please send me the logs when the error occurred.

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetwork.java:
```diff
@@ -172,7 +176,9 @@ public void train(final INDArray input, final int label) {
                                   final INDArray label) {
-    return backPropagateTo(0, activations, label);
+    // Process backpropagation to the second layer
+    // because the error the first layer generates, is not needed to generate a gradient for learning the first layer.
```
I'm still not satisfied with this explanation. This makes it sound like the first layer is always not learnable. Also, the words "the error the first layer generates" can be misunderstood if the reader doesn't know that weight updates are done by the next layer's back propagation call.

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/NeuralNetwork.java:
```diff
@@ -187,10 +193,15 @@ public void train(final INDArray input, final int label) {
-    return ArrayUtils.add(backPropagateFromTo(lastLayerIndex - 1, end, activations, error), error);
+    final INDArray error = layers[lastLayerIndex].backPropagate(label, activations[lastLayerIndex + 1], EMPTY);
+
+    if (lastLayerIndex == end) {
```
Is this for cases where the activationWithLoss layer is the only layer?

beomyeol replied:
This is the case that a neural network model has two layers and the last one is the `ActivationWithLossLayer`. I missed the case where the network model has only one layer.
In this case, the model cannot be trained since the gradients used for updating the layer should be calculated in the next layer's backpropagation. But, the next layer does not exist in the network.
I'll add a code for handling the case where the model has only one layer and comments about this.

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ActivationWithLossLayer.java:
```diff
@@ -0,0 +1,94 @@
+  @Override
+  public INDArray backPropagate(final INDArray label, final INDArray activation, final INDArray nextError) {
+    switch (lossFunction.toLowerCase()) {
+    case "cross-entropy":
```
The other names don't use hyphens; are we sure we want to use this naming convention just for loss functions?

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/conf/ActivationWithLossLayerConfigurationBuilder.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Configuration builder for activation with loss layer.
+ *
+ * The configuration that this builder generates is used to create an activation with layer instance.
```
"activation with **loss** layer instance"?

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ActivationWithLossLayer.java:
```diff
@@ -0,0 +1,94 @@
+import java.io.IOException;
+
+/**
+ * Loss Layer with activation function.
```
We should have comments saying that this layer is intended to be the final layer, in most cases.

jsjason commented on dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layers/ActivationLayer.java:
```diff
@@ -0,0 +1,78 @@
+import javax.inject.Inject;
+
+/**
+ * Activation Layer.
```
Shouldn't we include what this layer is actually doing, and various characteristics such as the number of inputs and outputs being the same?

jsjason commented:
The new ActivationWithLoss layer seems good. @beomyeol Please check my comments. I'll show you the logs later, offline.

beomyeol commented:
@jsjason Thanks for your review. I've addressed your comments. Please take another look.

jsjason commented:
@beomyeol The changes look great, but I have an extra question.
If I train a network with two FullyConnectedLayers and no ActivationWithLoss layers, I get an exception. But what I expected was that the application finishes correctly, although without any meaningful results. Is throwing an exception intended?

jsjason commented:
Here is the code I ran:
```Java
final Configuration nnConf = NeuralNetworkConfigurationBuilder.newConfigurationBuilder()
        .setBatchSize(1)
        .setStepsize(1e-2f)
        .setParameterProviderClass(LocalNeuralNetParameterProvider.class)
        .addLayerConfiguration(
            FullyConnectedLayerConfigurationBuilder.newConfigurationBuilder()
                .setNumInput(3)
                .setNumOutput(4)
                .setInitWeight(0.0001f)
                .setInitBias(0.0002f)
                .setRandomSeed(10)
                .build())
        .addLayerConfiguration(
            FullyConnectedLayerConfigurationBuilder.newConfigurationBuilder()
                .setNumInput(4)
                .setNumOutput(3)
                .setInitWeight(0.0001f)
                .setInitBias(0.0002f)
                .setRandomSeed(10)
                .build())
        .build();

    final Injector injector = Tang.Factory.getTang().newInjector(nnConf);
    neuralNetwork = injector.getInstance(NeuralNetwork.class);
    neuralNetwork.train(Nd4j.create(new float[]{1, 2, 3}), 1);
```

beomyeol commented:
@jsjason If the last layer is a fully connected layer, the backpropagation cannot work well since the fully connected layer needs the next layer's error. Because the error value that the last layer gets as the argument of the backpropagation cannot be defined. the neural network model uses the empty array as the argument of the last layer's backpropagation, which causes the exception. Thus, the last layer's backpropagation should generate an error without the next layer's error (e.g. `ActivationWithLossLayer`'s one).

jsjason commented:
@beomyeol Thanks. Another question: I know I'm asking you the same thing over and over again, but I'm confused yet again about excluding the first layer from back propagation. If we had n layers (not counting the input layer), shouldn't we have exactly n error vectors (matrices)? This code seems to output n-1 error vectors.

beomyeol commented:
@jsjason Yes, this code generate n-1 error vectors. The point that you may miss is that the last layer cannot be learnable. Let me suppose all layers except the last one are learnable. The network model can generate the parameter updates for these n-1 layers with n-1 error vectors. To be specific, the parameter updates for (n-1)-th layer is calculated with the error generated by n-th layer's backpropagation, the parameter update (n-2)-th layer is calculated with the error generated by (n-1)-th layer's one. That's why we can work well with n-1 error vectors. Please let me know if you are still confused.

jsjason commented:
@beomyeol Thanks, I understand what's going on in this PR now. I'll merge this after testing.
